lib/Dialect/Calyx/Export/CalyxEmitter.cpp

```diff
@@ -48,6 +44,54 @@ static constexpr std::string_view LBraceEndL() { return "{\n"; }
 static constexpr std::string_view RBraceEndL() { return "}\n"; }
 static constexpr std::string_view semicolonEndL() { return ";\n"; }
 
+/// A tracker to determine which libraries should be imported for a given
+/// program.
+struct ImportTracker {
+public:
+  /// Returns the list of library names used for in this program.
+  /// E.g. if `primitives/core.futil` is used, returns { "core" }.
+  SmallVector<StringRef> getLibraryNames(ProgramOp program) {
+    program.walk([&](ComponentOp component) {
+      for (auto &op : *component.getBody()) {
+        if (!op.hasTrait<Cell>() || isa<InstanceOp>(op))
+          // It is not a primitive.
+          continue;
+
+        StringRef opLibrary = getLibraryFor(&op);
+        if (llvm::any_of(usedLibraries,
```

Comment:
I suggest replacing `usedLibraries` with a set, and remove this manual lookup.

Reply:
Opted to use `SmallSet`, which remains a vector until you've exceeded the pre-allocated size. I think hashing is overkill otherwise.